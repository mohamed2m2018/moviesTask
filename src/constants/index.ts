export const colors = {
  grey: '#d8d8d8',
  green: '#59C14C',
  white: '#fff',
  black: '#000',
  background: '#F6F6F8',
};

export const testIds = {
  MOVIE_CARD_TITLE: 'MOVIE_CARD_TITLE',
  MOVIE_CARD_DATE: 'MOVIE_CARD_DATE',
  MOVIE_CARD_RATING: 'MOVIE_CARD_RATING',
  MOVIE_CARD_TOUCHABLE: 'MOVIE_CARD_TOUCHABLE',
  MOVIES_FLATLIST: 'MOVIES_FLATLIST',
};
