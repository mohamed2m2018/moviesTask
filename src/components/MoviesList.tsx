import React, {useCallback, useEffect} from 'react';
import {FlatList} from 'react-native';
import {useDispatch, useSelector} from 'react-redux';
import {useNavigation} from '@react-navigation/native';
import {getAllGeneres} from '../redux/actions/genresActions';
import MovieCard from './MovieCard';
import {GenresReducer, RootState} from '../redux/Model';
import styled from 'styled-components/native';
import {colors} from '../constants';
import {perfectHeight} from '../helpers/commonFunctions';

interface Props {
  movies: Object[];
  hasMore: boolean;
  fetching: boolean;
  action: Function;
  page: number;
}

const MoviesList = ({movies, hasMore, fetching, action, page}: Props) => {
  const genres = useSelector((state: RootState) => state.genres.genres);
  const dispatch = useDispatch();
  const navigation = useNavigation();
  useEffect(() => {
    if (movies?.length === 0) {
      dispatch(
        action?.({
          page,
        }),
      );
      dispatch(getAllGeneres());
    }
  }, [action, dispatch, movies?.length, page]);

  const renderItem = useCallback(
    ({item}) => {
      return (
        <MovieCard
          title={item?.title}
          date={item?.release_date}
          uri={item?.poster_path}
          rating={((item?.vote_average / 10) * 100).toFixed(0)}
          genres={item?.genre_ids?.map((id: string) => genres?.[id])}
          onPress={() =>
            navigation.navigate(
              'MovieDetails' as never,
              {
                title: item?.title,
                uri: item?.poster_path,
                overview: item?.overview,
                rating: ((item?.vote_average / 10) * 100).toFixed(0),
                genres: item?.genre_ids?.map(
                  (id: string) => genres?.[id as keyof GenresReducer],
                ),
                id: item?.id,
              } as never,
            )
          }
        />
      );
    },
    [genres, navigation],
  );

  const handleLoadMore = () => {
    if (hasMore && !fetching) {
      dispatch(action({page: page + 1}));
    }
  };
  return (
    <FlatList
      data={movies as []}
      renderItem={renderItem}
      keyExtractor={(item: {id: number}, index: number) =>
        item?.id + index?.toString()
      }
      onEndReachedThreshold={1}
      onEndReached={handleLoadMore}
      ListFooterComponent={() => {
        if (fetching) {
          return <ActivityIndicator size="large" color={colors.black} />;
        }
        return null;
      }}
    />
  );
};

const ActivityIndicator = styled.ActivityIndicator`
  margin-top: ${perfectHeight(10)}px;
`;

export default React.memo(MoviesList);
