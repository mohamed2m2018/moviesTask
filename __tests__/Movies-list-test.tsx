import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import {render} from '@testing-library/react-native';
import MoviesList from '../src/components/MoviesList';
import {testIds} from '../src/constants';

let props = {
  movies: [
    {
      adult: false,
      backdrop_path: '/kcvKEvTWqIGMjpVJdbDVRHdIt4C.jpg',
      genre_ids: [12, 28, 53],
      id: 370172,
      original_language: 'en',
      original_title: 'No Time to Die',
      overview:
        'Bond has left active service and is enjoying a tranquil life in Jamaica. His peace is short-lived when his old friend Felix Leiter from the CIA turns up asking for help. The mission to rescue a kidnapped scientist turns out to be far more treacherous than expected, leading Bond onto the trail of a mysterious villain armed with dangerous new technology.',
      popularity: 791.973,
      poster_path: '/iOPXG7q6ZqxSFp2iSY2CvgeRXyC.jpg',
      release_date: '2021-09-29',
      title: 'No Time to Die',
      video: false,
      vote_average: 7.3,
      vote_count: 225,
    },
    {
      adult: false,
      backdrop_path: '/5WPKkfVcMtQ0YOvSovbmxvNq3i1.jpg',
      genre_ids: [27, 53],
      id: 565028,
      original_language: 'en',
      original_title: 'Candyman',
      overview:
        'Anthony and his partner move into a loft in the now gentrified Cabrini-Green, and after a chance encounter with an old-timer exposes Anthony to the true story behind Candyman, he unknowingly opens a door to a complex past that unravels his own sanity and unleashes a terrifying wave of violence.',
      popularity: 581.797,
      poster_path: '/dqoshZPLNsXlC1qtz5n34raUyrE.jpg',
      release_date: '2021-08-25',
      title: 'Candyman',
      video: false,
      vote_average: 6.5,
      vote_count: 408,
    },
  ],
  hasMore: false,
  fetching: false,
  action: () => {},
  page: 1,
};

describe('Movie List', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  it('renders correctly', () => {
    const screenToBeSnapped = renderer
      .create(<MoviesList {...props} />)
      .toJSON();
    expect(screenToBeSnapped).toMatchSnapshot();
  });

  it('data rendered in flatlist are correct data from movies prop', async () => {
    const {getByTestId} = render(<MoviesList {...props} />);
    const list = getByTestId(testIds.MOVIES_FLATLIST);
    expect(list.props.data).toStrictEqual(props?.movies);
  });
});
