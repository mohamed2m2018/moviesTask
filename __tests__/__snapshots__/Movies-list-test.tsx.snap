// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie List renders correctly 1`] = `
<RCTScrollView
  ListFooterComponent={[Function]}
  data={
    Array [
      Object {
        "adult": false,
        "backdrop_path": "/kcvKEvTWqIGMjpVJdbDVRHdIt4C.jpg",
        "genre_ids": Array [
          12,
          28,
          53,
        ],
        "id": 370172,
        "original_language": "en",
        "original_title": "No Time to Die",
        "overview": "Bond has left active service and is enjoying a tranquil life in Jamaica. His peace is short-lived when his old friend Felix Leiter from the CIA turns up asking for help. The mission to rescue a kidnapped scientist turns out to be far more treacherous than expected, leading Bond onto the trail of a mysterious villain armed with dangerous new technology.",
        "popularity": 791.973,
        "poster_path": "/iOPXG7q6ZqxSFp2iSY2CvgeRXyC.jpg",
        "release_date": "2021-09-29",
        "title": "No Time to Die",
        "video": false,
        "vote_average": 7.3,
        "vote_count": 225,
      },
      Object {
        "adult": false,
        "backdrop_path": "/5WPKkfVcMtQ0YOvSovbmxvNq3i1.jpg",
        "genre_ids": Array [
          27,
          53,
        ],
        "id": 565028,
        "original_language": "en",
        "original_title": "Candyman",
        "overview": "Anthony and his partner move into a loft in the now gentrified Cabrini-Green, and after a chance encounter with an old-timer exposes Anthony to the true story behind Candyman, he unknowingly opens a door to a complex past that unravels his own sanity and unleashes a terrifying wave of violence.",
        "popularity": 581.797,
        "poster_path": "/dqoshZPLNsXlC1qtz5n34raUyrE.jpg",
        "release_date": "2021-08-25",
        "title": "Candyman",
        "video": false,
        "vote_average": 6.5,
        "vote_count": 408,
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onEndReachedThreshold={1}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  testID="MOVIES_FLATLIST"
  viewabilityConfigCallbackPairs={Array []}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderRadius": 10,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginVertical": 20.523076923076925,
            "opacity": 1,
            "paddingBottom": 20.523076923076925,
            "paddingLeft": 20.523076923076925,
            "paddingRight": 20.523076923076925,
            "paddingTop": 20.523076923076925,
            "width": 700,
          }
        }
        testID="MOVIE_CARD_TOUCHABLE"
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "column",
                "width": 240,
              },
            ]
          }
        >
          <Image
            resizeMode="cover"
            source={
              Object {
                "uri": "https://image.tmdb.org/t/p/original//iOPXG7q6ZqxSFp2iSY2CvgeRXyC.jpg",
              }
            }
            style={
              Array [
                Object {
                  "borderRadius": 10,
                  "height": 246.27692307692305,
                  "width": 200,
                },
              ]
            }
          />
        </View>
        <View
          style={
            Array [
              Object {
                "flexDirection": "column",
                "width": 320,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 20,
                  "fontWeight": "bold",
                  "marginBottom": 10,
                },
              ]
            }
            testID="MOVIE_CARD_TITLE"
          >
            No Time to Die
          </Text>
          <Text
            style={
              Array [
                Object {
                  "fontSize": 15,
                  "marginBottom": 20,
                },
              ]
            }
            testID="MOVIE_CARD_DATE"
          >
            2021-09-29
          </Text>
          <View
            style={
              Array [
                Object {
                  "flexDirection": "row",
                  "flexWrap": "wrap",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "#d8d8d8",
                    "borderRadius": 10,
                    "marginBottom": 10.261538461538462,
                    "marginRight": 10,
                    "paddingHorizontal": 10,
                  },
                ]
              }
            >
              <Text />
            </View>
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "#d8d8d8",
                    "borderRadius": 10,
                    "marginBottom": 10.261538461538462,
                    "marginRight": 10,
                    "paddingHorizontal": 10,
                  },
                ]
              }
            >
              <Text />
            </View>
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "#d8d8d8",
                    "borderRadius": 10,
                    "marginBottom": 10.261538461538462,
                    "marginRight": 10,
                    "paddingHorizontal": 10,
                  },
                ]
              }
            >
              <Text />
            </View>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "flexDirection": "column",
                "justifyContent": "flex-end",
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#73F340",
                  "fontSize": 20,
                  "fontWeight": "bold",
                },
              ]
            }
            testID="MOVIE_CARD_RATING"
          >
            73
            %
          </Text>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderRadius": 10,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginVertical": 20.523076923076925,
            "opacity": 1,
            "paddingBottom": 20.523076923076925,
            "paddingLeft": 20.523076923076925,
            "paddingRight": 20.523076923076925,
            "paddingTop": 20.523076923076925,
            "width": 700,
          }
        }
        testID="MOVIE_CARD_TOUCHABLE"
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "column",
                "width": 240,
              },
            ]
          }
        >
          <Image
            resizeMode="cover"
            source={
              Object {
                "uri": "https://image.tmdb.org/t/p/original//dqoshZPLNsXlC1qtz5n34raUyrE.jpg",
              }
            }
            style={
              Array [
                Object {
                  "borderRadius": 10,
                  "height": 246.27692307692305,
                  "width": 200,
                },
              ]
            }
          />
        </View>
        <View
          style={
            Array [
              Object {
                "flexDirection": "column",
                "width": 320,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 20,
                  "fontWeight": "bold",
                  "marginBottom": 10,
                },
              ]
            }
            testID="MOVIE_CARD_TITLE"
          >
            Candyman
          </Text>
          <Text
            style={
              Array [
                Object {
                  "fontSize": 15,
                  "marginBottom": 20,
                },
              ]
            }
            testID="MOVIE_CARD_DATE"
          >
            2021-08-25
          </Text>
          <View
            style={
              Array [
                Object {
                  "flexDirection": "row",
                  "flexWrap": "wrap",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "#d8d8d8",
                    "borderRadius": 10,
                    "marginBottom": 10.261538461538462,
                    "marginRight": 10,
                    "paddingHorizontal": 10,
                  },
                ]
              }
            >
              <Text />
            </View>
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "#d8d8d8",
                    "borderRadius": 10,
                    "marginBottom": 10.261538461538462,
                    "marginRight": 10,
                    "paddingHorizontal": 10,
                  },
                ]
              }
            >
              <Text />
            </View>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "flexDirection": "column",
                "justifyContent": "flex-end",
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#73F340",
                  "fontSize": 20,
                  "fontWeight": "bold",
                },
              ]
            }
            testID="MOVIE_CARD_RATING"
          >
            65
            %
          </Text>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
    />
  </View>
</RCTScrollView>
`;
